bpo-14243: Optionally delete NamedTemporaryFile on content manager exit but not on file close

Ev2geny opened the pull request:
This fixes this issue: https://bugs.python.org/issue14243

```tempfile.NamedTemporaryFile``` is too hard to use portably when you need to open the file by name after writing it.  To do that, you need to close the file first (on Windows), which means you have to pass ```delete=False```, which in turn means that you get no help in cleaning up the actual file resource,

Hence at the moment there is no out of the box solution to use  tempfile.NamedTemporaryFile on Windows in such scenario (which is often used in unit testing):

* in test module:
1) create and open temporary file
2) write data to it
3) pass name of the temporary file to the operational code

* In operational code, being tested
1) open file, using name of the temporary file 
2) read data from this temporary file 

In this Pull Request the issue is solved by adding an additional optional argument to ```NamedTemporaryFile() 'delete_on_close' ``` (default is True). It works in combination with already existing argument ```'delete'```, and determines whether created temporary file gets deleted on close (which until this change was the only functionality available and it remains default now).

If  ```'delete' = True``` and  ```'delete_on_close' = False``` then temporary file gets deleted on context manager exit only (which is a new functionality).

If ```'delete'``` is not true, then value of '```delete_on_close'``` is ignored.

So, the change shall be fully backwards compatible.

In the bug discussion such option was discussed by [r.david.murray ](https://bugs.python.org/msg157947) but mainly by [eryksun ](https://bugs.python.org/msg376656)

I have also added several unit tests to test new functionality in different combinations


<!-- issue-number: [bpo-14243](https://bugs.python.org/issue14243) -->
https://bugs.python.org/issue14243
<!-- /issue-number -->

Ev2geny committed:
inial update. Tests do not work yet

Ev2geny committed:
delete_on_close flag is added to NamedTemporaryFile functionality

Ev2geny committed:
possibility for user to delete temporary file is added without content manager complaining

Ev2geny committed:
typo in test name is fixed

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Ev2geny committed:
whatsnew and News file is updated

Ev2geny committed:
bro-14243: update to NEWS.d

zooba commented on Lib/test/test_tempfile.py:
```diff
@@ -903,6 +903,65 @@ def test_del_on_close(self):
+            # setting delete_on_close = True to test, that this does not have
+            # an effect, if delete = False
+            with tempfile.NamedTemporaryFile(dir=dir, delete = False,
+                                             delete_on_close = True) as f:
```
```suggestion
            with tempfile.NamedTemporaryFile(dir=dir, delete=False,
                                             delete_on_close=True) as f:
```

zooba commented on Misc/NEWS.d/next/Library/2020-09-28-04-56-04.bpo-14243.YECnxv.rst:
```diff
@@ -0,0 +1,16 @@
+At the moment  ``NamedTemporaryFile`` is too hard to use portably when you
```
This is way more details than we want here - this will end up in a list with 1000 other changes.

Most of this should move to specific documentation, and then this entry can link to it with ``:class:`NamedTemporaryFile` ``

Ev2geny replied:
This is fixed. The spesific documentation is updated and linked to with ` :class:`tempfile.NamedTemporaryFile``

zooba commented on Doc/whatsnew/3.10.rst:
```diff
@@ -93,6 +93,8 @@ Other Language Changes
   :meth:`~object.__index__` method).
   (Contributed by Serhiy Storchaka in :issue:`37999`.)
 
+* XXX Describe changes to NamedTemporaryFile functionality. Re 2020-09-28-04-56-04.bpo-14243.YECnxv.rst
```
Needs to move to `3.12.rst`, and can be about half the length of what's currently in the NEWS file

Ev2geny replied:
Implemented. Whats new is moved to 3.12.rst and only 2 lines now

zooba reviewed (changes requested):
Code change looks good, but we need to get the docs in order.

Ev2geny committed:
Update Lib/test/test_tempfile.py

Commiting changes, suggested by @zooba

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

Ev2geny committed:
Doc/whatsnew/3.10.rst is restored from the main. changes will no go to 3.12 rst

Ev2geny committed:
Documentation update

Ev2geny committed:
Minor documentation update

Ev2geny committed:
Training whitespaces removed in Doc/library/tempfile.rst

Ev2geny committed:
New line is added to the end of the news file

AA-Turner commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,35 @@ The module defines the following user-callable items:
+
+   * The file is guaranteed to have a visible name in the file system (on Unix, the directory entry is not unlinked).
+
+   * The *delete_on_close* parameter is added.
```
```suggestion
   * There is more granularity in the deletion behaviour of the file
     (see ``delete_on_close`` below)
```

AA-Turner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -79,6 +79,7 @@ New Features
 Other Language Changes
 ======================
 
+* XXX The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :issue:`14243`.)
```
This should go in Improved Modules -> tempfile
```suggestion
* The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :gh:`22431`.)
```

AA-Turner replied:
@Ev2geny -- you updated the text but you need to move it to the section I mentioned (see https://github.com/python/cpython/blob/5893b5db98b38b17750c0572c7209774a5034898/Doc/whatsnew/3.9.rst?plain=1#L327-L332) for what the section should look like -- you'll need to add `tempfile` as it's the first improvement listed in whatsnew.

A

Ev2geny replied:
@AA-Turner , thanks. i have fixed it now.

Ev2geny committed:
Update Doc/library/tempfile.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

Ev2geny committed:
Update Doc/whatsnew/3.12.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

zooba commented on Doc/whatsnew/3.12.rst:
```diff
@@ -79,6 +79,7 @@ New Features
 Other Language Changes
 ======================
 
+* The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :gh:`22431`.)
```
```suggestion
* The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :issue:`14243`.)
```

We typically refer to the issue number, not the pull request. That should also get around the (incorrect) error here.

AA-Turner replied:
Sorry, copy and paste error - I meant to suggest updating the bpo `:issue:` role to GitHub -- the number should be #58451

```suggestion
* The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :gh:`58451`.)
```

A

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,36 @@ The module defines the following user-callable items:
+   The name of the temporary file can be retrieved from the :attr:`name`
+   attribute of the returned file-like object. Whether the name can be used to
+   open the file a second time, while the named temporary file is still open,
+   varies across platforms (it can be so used on Unix; it cannot on Windows).
```
In Windows, the file can be opened again if `delete_on_close` is false or if the open shares delete access (e.g. by calling `CreateFileW()` directly with `FILE_SHARE_DELETE`, or calling `os.open()` with `O_TEMPORARY`). If `delete_on_close` is false, and the file is opened again *without* sharing delete access (e.g. via builtin `open()`), then the second open *must* be closed before exiting the context manager, else the `unlink()` call on context manager exit will fail with a `PermissionError` due to an OS sharing-violation error (32).

Ev2geny replied:
@eryksun I have implemented your comments. Can you please look at this once again?

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,36 @@ The module defines the following user-callable items:
+
+   If *delete* is true (the default) and *delete_on_close* is true (the
+   default), the file is deleted as soon as it is closed. If *delete* is true
+   and *delete_on_close* is false, the file is deleted on context manager exit
```
In Windows, if `delete_on_close` is false, and the file is created in a directory for which the user lacks delete access (either directly for files as an inherited rule that grants `DELETE` access, or at the directory level via `FILE_DELETE_CHILD` access), then the `unlink()` call on exit of the context manager will fail with a `PermissionError` due to an OS access-denied error (5). This cannot happen when `delete_on_close` is true because `DELETE` access is implicitly requested when `CreateFileW()` is called with `FILE_FLAG_DELETE_ON_CLOSE`, so the open immediately fails if the caller lacks delete access.

Ev2geny replied:
@eryksun I have implemented your comments. Thanks!
I have also slightly rearranged information. I think the flow is more logical now

Ev2geny committed:
Update Doc/whatsnew/3.12.rst

changing issue number to github issue number in documentation

Co-authored-by: Steve Dower <steve.dower@microsoft.com>

Ev2geny committed:
Documentation in tempfile.rst is updated based on the feedback from @eryksun

Ev2geny committed:
Merge branch 'fix-issue-14243' of https://github.com/Ev2geny/cpython into fix-issue-14243

Ev2geny committed:
Information in whatsnew is moved to the section Improved modules based on the feedback from AA-Turner

AA-Turner commented on Doc/whatsnew/3.12.rst:
```diff
@@ -90,6 +90,10 @@ New Modules
+tempfile
+--------
+
+The :class:`tempfile.NamedTemporaryFile` has a new optional parameter *delete_on_close* (Contributed by Evgeny Zorin in :gh:`58451`.)
```
Wrap lines at 70-80 chars
```suggestion
The :class:`tempfile.NamedTemporaryFile` has a new optional parameter 
*delete_on_close* (Contributed by Evgeny Zorin in :gh:`58451`.)
```

Ev2geny committed:
Update Doc/whatsnew/3.12.rst

Wrap lines at 70-80 chars in the Doc/whatsnew/3.12.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,57 @@ The module defines the following user-callable items:
+     delete access (e.g. via builtin :func:`open`), then the second open must be
+     closed before exiting the context manager, else the :func:`os.unlink` call on
+     context manager exit will fail with a ``PermissionError`` due to an OS
+     sharing-violation error (32).
```
I think just the Python `os.open` example suffices, and the Windows error code is too much information:
 
```suggestion
   * In Windows, the file can be opened again if *delete_on_close* is false or
     if the open shares delete access (e.g. by calling :func:`os.open` with the
     flag ``O_TEMPORARY``).  If *delete_on_close* is false, and the file is
     opened again without sharing delete access (e.g. via builtin :func:`open`),
     then the second open must be closed before exiting the context manager,
     else the :func:`os.unlink` call on context manager exit will fail with a
     ``PermissionError``.
```

Ev2geny replied:
OK, thanks.

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,57 @@ The module defines the following user-callable items:
+   access-denied error (5). This cannot happen when *delete_on_close* is true
+   because ``DELETE`` access is implicitly requested when ``CreateFileW()`` is
+   called with ``FILE_FLAG_DELETE_ON_CLOSE``, so the open immediately fails if
+   the caller lacks delete access.
```
Again, simplify since there's no need to overwhelm the reader with too much information:

```suggestion
   In Windows, if *delete_on_close* is false, and the file is created in a
   directory for which the user lacks delete access, then the :func:`os.unlink`
   call on exit of the context manager will fail with a ``PermissionError``.
   This cannot happen when *delete_on_close* is true because delete access is
   requested by the open, which fails immediately if the requested access is
   not granted.
```

Ev2geny committed:
Update Doc/library/tempfile.rst

Update to documentation (Doc/library/tempfile.rst) based on the review of eryksun

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Update Doc/library/tempfile.rst

Update to documentation (Doc/library/tempfile.rst) based on the review of eryksun

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
wrapping at 80 characters is implemented on the Doc/library/tempfile.rs

merwok commented on Doc/whatsnew/3.12.rst:
```diff
@@ -90,6 +90,11 @@ New Modules
+tempfile
+--------
+
+The :class:`tempfile.NamedTemporaryFile` has a new optional parameter 
```
```suggestion
The :class:`tempfile.NamedTemporaryFile` class has a new optional parameter 
```

merwok commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,54 @@ The module defines the following user-callable items:
+   Whether the name of the temporary file can be used to open the file a second
+   time, while the named temporary file is still open, varies across platforms:
+
+   * It can be so used on Unix.
```
I don’t understand `so used`

merwok commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,54 @@ The module defines the following user-callable items:
+   To use the name of the temporary file to open the closed file second time,
+   either make sure not to delete the file upon closure (set the *delete*
+   parameter to be false) or, in case the temporary file is created in a
+   :keyword:`with` statement, set the *delete_on_close* to be false. The latter
```
Say either:

> set *delete_on_close* to false

or:

> set the *delete_on_close* parameter to false

Ev2geny replied:
Thanks, this is implemented

merwok commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,54 @@ The module defines the following user-callable items:
+
+   * It can be so used on Unix.
+
+   * In Windows, the file can be opened again if *delete_on_close* is false or
```
On Windows

(same tiny issue in other places in the file, please change all)

Ev2geny replied:
Fixed

merwok commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,54 @@ The module defines the following user-callable items:
+
+   In Windows, if *delete_on_close* is false, and the file is created in a
+   directory for which the user lacks delete access, then the :func:`os.unlink`
+   call on exit of the context manager will fail with a ``PermissionError``.
```
Could `PermissionError` be a link?  (using `:exc:` role if that exists)

Ev2geny replied:
Thank you! Implemented!

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,54 @@ The module defines the following user-callable items:
+   parameter to be false) or, in case the temporary file is created in a
+   :keyword:`with` statement, set the *delete_on_close* to be false. The latter
+   approach is recommended as it provides assistance in automatic cleaning of
+   the temporary file upon the context manager exit.
```
Since only POSIX and Windows are discussed, and it's mostly about Windows, the bullet points and subsequent paragraph could be folded into a single paragraph. For example:

```suggestion
   While the named temporary file is open, the file can always be opened again
   on POSIX. On Windows, it can be opened again if *delete* is false, or if
   *delete_on_close* is false, or if the additional open shares delete access
   (e.g. by calling :func:`os.open` with the flag ``O_TEMPORARY``).  On
   Windows, if *delete* is true and *delete_on_close* is false, additional
   opens that do not share delete access (e.g. via builtin :func:`open`) must
   be closed before exiting the context manager, else the :func:`os.unlink`
   call on context manager exit will fail with a ``PermissionError``.
```

Ev2geny replied:
@eryksun , I have implemented you suggestion, but then restored the (probably accidently deleted) paragraph about usage of the name of the file to open it second time. This is the main reason this PR was created, I think I have to leave it in.

```
   To use the name of the temporary file to open the closed file second time,
   either make sure not to delete the file upon closure (set the *delete*
   parameter to be false) or, in case the temporary file is created in a
   :keyword:`with` statement, set the *delete_on_close* to be false. The latter
   approach is recommended as it provides assistance in automatic cleaning of
   the temporary file upon the context manager exit.
```

Ev2geny committed:
Update Doc/whatsnew/3.12.rst

Co-authored-by: Éric <merwok@netwok.org>

Ev2geny committed:
Update Doc/library/tempfile.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Restored (probably accidentry removed by @eryksun's suggestion) paragraph in the documentation text about the use of the name of the temporary file to open the closed file second time

Ev2geny committed:
Additional minor changes to documentation, based on suggestions from @merwok

eryksun commented on Lib/tempfile.py:
```diff
@@ -521,17 +533,23 @@ def __iter__(self):
-    'delete' -- whether the file is deleted on close (default True).
+    'delete' -- whether the file is deleted at the end.  When exactly file gets
+       deleted (either on close or on context manager exit) is determined by
+       parameter delete_on_close.  (default True).
```
Trying to cut down on the verbosity:

```suggestion
    'delete' -- whether the file is automatically deleted (default True).
    'delete_on_close' -- if 'delete', whether the file is deleted on close
       or on context exit (default True).
```

eryksun commented on Lib/tempfile.py:
```diff
@@ -521,17 +533,23 @@ def __iter__(self):
     The file is created as mkstemp() would do it.
+    'delete_on_close' -- if 'delete', then determines whether file gets deleted
+       on close. Otherwise it gets deleted on context manager exit
+       (default True)
```
The explanation of this parameter shouldn't come after "the file is created as mkstemp() would do it". I think it's better if it immediately follows the `delete` parameter.  
```suggestion
```

eryksun commented on Lib/tempfile.py:
```diff
@@ -500,6 +503,15 @@ def __enter__(self):
     def __exit__(self, exc, value, tb):
         result = self.file.__exit__(exc, value, tb)
         self.close()
+        # if file is to be deleted, bot not on closure, deleting it now
```
```suggestion
        # If the file is to be deleted, but not on close, delete it now.
```

eryksun commented on Lib/tempfile.py:
```diff
@@ -500,6 +503,15 @@ def __enter__(self):
+            try:
+                _os.unlink(self.name)
+            # It shall be Ok to ignore FileNotFoundError, because user may
+            # have deleted the file already before content manager came to it
```
```suggestion
            # It is okay to ignore FileNotFoundError. The file may have
            # been deleted already.
```

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Update Lib/tempfile.py

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Merge branch 'fix-issue-14243' of https://github.com/Ev2geny/cpython into fix-issue-14243

eryksun commented on Lib/tempfile.py:
```diff
@@ -500,6 +503,15 @@ def __enter__(self):
     def __exit__(self, exc, value, tb):
         result = self.file.__exit__(exc, value, tb)
         self.close()
+        # If the file is to be deleted, but not on close, delete it now.
+        if self.delete and not self.delete_on_close:
+            try:
+                _os.unlink(self.name)
+            # It is okay to ignore FileNotFoundError. The file may have
+            # been deleted already.
+            except FileNotFoundError:
+                pass
+
```
A case hasn't been covered. If `delete` is true and `delete_on_close` is false, but the file is never used in a `with` context, then it won't be deleted automatically when the instance is finalized. A `__del__()` method is needed for this case.

Ev2geny replied:
@eryksun , I am just trying to figure out, whether this is the issue introduced by my changes or this is a separate issue, which existed before?

eryksun replied:
It's a new case to cover. Before there was just `delete` and `delete_on_close` was implied. On POSIX, when `delete_on_close` is true, finalizing the `_TemporaryFileCloser` instance deletes the file. On Windows, the OS deletes the file. OTOH, when `delete` is true and `delete_on_close` is false, the file has to be deleted when the `_TemporaryFileWrapper` instance is finalized, in `__del__()`.

Ev2geny replied:
OK, I think I understand now. Also the new test case needs to be created to cover this, I think. Correct?

Ev2geny replied:
This is now implemented in the new [PR 97015](https://github.com/python/cpython/pull/97015)

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,50 @@ The module defines the following user-callable items:
-   attribute is the underlying true file object. This file-like object can
-   be used in a :keyword:`with` statement, just like a normal file.
+   attribute is the underlying true file object. This file-like object can be
+   used in a :keyword:`with` statement, just like a normal file.The name of the
```
```suggestion
   used in a :keyword:`with` statement, just like a normal file.  The name of the
```

eryksun commented on Doc/library/tempfile.rst:
```diff
@@ -75,20 +75,50 @@ The module defines the following user-callable items:
+   be closed before exiting the context manager, else the :func:`os.unlink`
+   call on context manager exit will fail with a :exc:`PermissionError`.
+
+   To use the name of the temporary file to open the closed file second time,
+   either make sure not to delete the file upon closure (set the *delete*
+   parameter to be false) or, in case the temporary file is created in a
+   :keyword:`with` statement, set the *delete_on_close* parameter to be false.
+   The latter approach is recommended as it provides assistance in automatic
+   cleaning of the temporary file upon the context manager exit.
```
The previous paragraphs explain the basic operation of the `delete` and `delete_on_close` parameters and that the file can be opened again if `delete` or `delete_on_close` is false. I was trying to avoid restating this because it's getting kind of long-winded.

Ev2geny replied:
I think one can work out this information from the previous pragagraphs, but from the not so advanced user  point of view (like myself) I would want to have this paragraph as well, as it describes arguably the most frequent use of the named temporary file: write some data, close it and then use its name to open it further.

Ev2geny committed:
Update Doc/library/tempfile.rst

Co-authored-by: Eryk Sun <eryksun@gmail.com>

Ev2geny committed:
Merged with master and resolved merge conflicsts in Doc/whatsnew/3.12.rst and in Lib/tempfile.py

Ev2geny committed:
Revert "Merged with master and resolved merge conflicsts in Doc/whatsnew/3.12.rst and in Lib/tempfile.py"

This reverts commit 07076e83126f246d1480418f46071d8e2cdedf80.
